core/src/main/java/software/amazon/awssdk/core/signer/Signer.java

```diff
@@ -13,13 +13,12 @@
  * permissions and limitations under the License.
  */
 
-package software.amazon.awssdk.core.runtime.auth;
+package software.amazon.awssdk.core.signer;
 
-import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
 
-@SdkProtectedApi
-public abstract class SignerProvider {
+public interface Signer {
 
-    public abstract Signer getSigner(SignerProviderContext context);
-
-}
+    SdkHttpFullRequest sign(SdkHttpFullRequest request, ExecutionAttributes executionAttributes);
```

Comment:
I'm worried that it's tough for someone to use the signer directly, because it's not obvious what they need to set in the `ExecutionAttributes` (partly for the reason we like it - it varies from signer to signer).

Is there a way to let us still use `ExecutionAttributes`, but also standardize that signers should have an explicit "this is what we need" interface for people using the signers directly?

Something like:
```Java
// Using it directly...
SdkHttpFullRequest httpRequest = ...;
Signer<AwsV4SigningRequest> signer = AwsV4Signer.create();
AwsV4SigningRequest signingRequest = AwsV4SigningRequest.builder()
                                                        .awsCredentials(...)
                                                        .region(...)
                                                        .serviceName(...)
                                                        .doubleUrlEncode(...)
                                                        .build();
SdkHttpFullRequest signedRequest = signer.sign(signingRequest, httpRequest);

// Using it from core...
SdkHttpFullRequest httpRequest = ...;
Signer<?> signer = ...;
ExecutionAttributes executionAttributes = ...;
SdkHttpFullRequest signedRequest = signer.sign(executionAttributes, httpRequest);
```

Reply:
>>  because it's not obvious what they need to set in the ExecutionAttributes

I agree its not obvious but we can document the required params for each signer implementation. So advanced users who want to use signers directly can know what params are required.

>> signer.sign(signingRequest, httpRequest);

This is confusing to me why we are passing in 2 requests

Reply:
I just renamed `AwsV4SigningParams` into `AwsV4SigningRequest`.

If the usability studies have taught me anything, no one reads documentation. I'd much prefer they be able to figure it out on their own!

Reply:
One other downside is signing parameters that might change. For example, I am setting S3ExecutionAttributes.ENABLE_CHUNKED_ENCODING in EnableChunkedEncodingInterceptor class. To support the new sign method, this parameter needs to be modeled too. In future, s3 has new requirements for payload signing, we might need to deprecate/add new params. Are we okay with modeling every optional parameter we use for signing?